Adds initial code

johnbradley opened the pull request:
Adds `dm_formatter.py` script containing @thibaulttabarin's Minnows specific metadata transformation code.
Adds a `Dockerfile` and GitHub Action config to build on release.
Adds a test script to exercise the metadata transformation code and GitHub Action config to run the tests.

johnbradley committed:
Add initial thibaulttabarin code

Add function written by thibaulttabarin from the drexel_metadata
repo.
See https://github.com/hdr-bgnn/drexel_metadata/blob/9d616a6b32a2b2786455e05a09dc799498a38af7/gen_metadata.py#L1002-L1058

Co-authored-by: thibaulttabarin <tabarin@battelleecology.org>

johnbradley committed:
Add command line options to dm_formatter.py

johnbradley committed:
Add Dockerfile

johnbradley committed:
Update README.md

johnbradley committed:
Add logic to build a ghcr.io container on release

Based on https://docs.github.com/en/packages/managing-github-packages-using-github-actions-workflows/publishing-and-installing-a-package-with-github-actions#publishing-a-package-using-an-action

johnbradley committed:
Add unit tests

Also adds a github action to run tests based on:
https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-python#using-the-python-starter-workflow

hlapp commented on dm_formatter.py:
```diff
@@ -0,0 +1,84 @@
+    """
+    Reformat and reduce the size of the result dictionary. 
+    Collect only the data necessary for BGNN minnow project. The new format matches the 
+    BGNN_metadata version. Therefore some of the value not calcualted in drexel version are by 
```
```suggestion
    BGNN_metadata version. Therefore some of the values not calculated in drexel version are by 
```

hlapp commented on dm_formatter.py:
```diff
@@ -0,0 +1,84 @@
+    Reformat and reduce the size of the result dictionary. 
+    Collect only the data necessary for BGNN minnow project. The new format matches the 
+    BGNN_metadata version. Therefore some of the value not calcualted in drexel version are by 
+    defaulset to "None". 
```
```suggestion
    default set to "None". 
```

hlapp commented on dm_formatter.py:
```diff
@@ -0,0 +1,84 @@
+    # Fish metadata
+    fish_num = first_value['fish_count']
+    fish_bbox = first_value['fish'][0]['bbox']
+    pixel_analysis = False if first_value['fish'][0]['pixel_analysis_failed'] else True
```
Wouldn't it be simpler to just use this assignment?
```suggestion
    pixel_analysis = !first_value['fish'][0]['pixel_analysis_failed']
```

johnbradley replied:
FYI: This should have been `pixel_analysis = not first_value['fish'][0]['pixel_analysis_failed']`.

hlapp commented on Dockerfile:
```diff
@@ -0,0 +1,6 @@
+WORKDIR /pipeline
+
+COPY dm_formatter.py /pipeline
+CMD echo "python dm_formatter.py [input] [output]"
```
This makes it look like input and output are optional, which it doesn't seem they actually are, and it also differs from the Usage section in the README.

hlapp commented on dm_formatter.py:
```diff
@@ -0,0 +1,84 @@
+    foreground_mean = first_value['fish'][0]['foreground']['mean']
+    background_mean = first_value['fish'][0]['background']['mean']
+        
+    dict_fish = {'fish_num': fish_num,"bbox":fish_bbox, 
```
```suggestion
    dict_fish = {'fish_num': fish_num,"bbox": fish_bbox, 
```

hlapp commented on dm_formatter.py:
```diff
@@ -0,0 +1,84 @@
+    background_mean = first_value['fish'][0]['background']['mean']
+        
+    dict_fish = {'fish_num': fish_num,"bbox":fish_bbox, 
+                 'pixel_analysis':pixel_analysis, 'rescale':"None", 
```
```suggestion
                 'pixel_analysis': pixel_analysis, 'rescale': "None", 
```

hlapp commented on dm_formatter.py:
```diff
@@ -0,0 +1,84 @@
+        
+    dict_fish = {'fish_num': fish_num,"bbox":fish_bbox, 
+                 'pixel_analysis':pixel_analysis, 'rescale':"None", 
+                 'eye_bbox': "None", 'eye_center':eye_center , 'angle_degree': "None",
```
```suggestion
                 'eye_bbox': "None", 'eye_center': eye_center , 'angle_degree': "None",
```

hlapp commented on dm_formatter.py:
```diff
@@ -0,0 +1,84 @@
+    dict_fish = {'fish_num': fish_num,"bbox":fish_bbox, 
+                 'pixel_analysis':pixel_analysis, 'rescale':"None", 
+                 'eye_bbox': "None", 'eye_center':eye_center , 'angle_degree': "None",
+                 'eye_direction':eye_direction, 'foreground_mean':round(foreground_mean,2), 
```
```suggestion
                 'eye_direction': eye_direction, 'foreground_mean': round(foreground_mean,2), 
```

hlapp commented on dm_formatter.py:
```diff
@@ -0,0 +1,84 @@
+                 'pixel_analysis':pixel_analysis, 'rescale':"None", 
+                 'eye_bbox': "None", 'eye_center':eye_center , 'angle_degree': "None",
+                 'eye_direction':eye_direction, 'foreground_mean':round(foreground_mean,2), 
+                 'background_mean':round(background_mean,2)}
```
```suggestion
                 'background_mean': round(background_mean,2)}
```

hlapp reviewed (approved):
Comments and suggestions are all very minor, so I'm approving in advance.

johnbradley committed:
Add minor fixes

Adds various fixes suggested in PR #1.

Co-authored-by: Hilmar Lapp <hlapp@drycafe.net>

johnbradley commented:
@thibaulttabarin Do you have any feedback for this PR?

thibaulttabarin commented:
> 

I didn't have time to do it throughly, it looks good but I want to test everything and i didn't have to finish to look at the drexel_metadata.

thibaulttabarin reviewed (changes requested):
add 
#!/usr/bin/env python3

thibaulttabarin reviewed (changes requested):
add 

# ADD scripts in /pipeline to the PATH
ENV PATH="/pipeline:${PATH}"

johnbradley committed:
Simplify Dockerfile usage

Adds dm_formatter.py to PATH so users can run it directly.
Also adds metadata to the Docker container.

johnbradley commented:
@thibaulttabarin I added changes to put dm_formatter.py into the PATH and added the shebang.
If you have a metadata file named `INHS_FISH_74711_metadata.json` in the current directory you should be able to test out the changes like so:
```
docker build -t dmf .
docker run -v $(pwd):/data -it dmf dm_formatter.py /data/INHS_FISH_74711_metadata.json /data/result.json
```
